Fix safe zone id reuse for properties

burqen opened the pull request:
Make sure termination of transactions is checked during lazy
property evaluation by pushing down the statement is open assertion
to cursor level.

During slave pull updates the `TransactionBatchCommitter` will mark transactions falling outside of safe zone for termination. 

Consider this scenario:
```
try ( Transaction tx = db.beginTx )
{
    Object property = db.getNodeById( id ).getProperty( propKey );
    // Act on information from property
    tx.success();
}
```
Until we do `tx.success` we don't know if our transaction has been terminated or not and we could potentially act on inconsistent data (from the dynamic store).

The current workaround for this is to assert that statement is still open before returning. We do this for 
`nodeGetProperty( long nodeId, int propertyKeyId )`
`relationshipGetProperty( long nodeId, int propertyKeyId )`
`nodeGetPropertyKeys( long nodeId )`
`relationshipGetPropertyKeys( long nodeId )`
but not for any other `getProperty/Properties`.
The updated `TerminationOfSlavesDuringPullUpdatesTest` shows that this is a problem and this PR tries to solve it by pushing down the assert open check to the cursor level.

Commit:
Fix safe zone id reuse for properties

Make sure termination of transactions is checked during lazy
property evaluation by pushing down the statement is open assertion
to cursor level.

burqen commented:
Forward merge here: https://github.com/burqen/neo4j/commit/8f12f2bb7198b6d7ae83b9f8cf620d623a126759

chrisvest commented:
@burqen Does this need a change log label?

burqen commented:
This is essentially a bug fix and as far as I know no one has run into the problem, so my first thought is 'no'.

Commit:
Fix checkstyle

tinwelint commented:
I'd like a better description of exactly what this PR fixes and what the problem was.

tinwelint reviewed (changes requested):
I'd like a better description of exactly what this PR fixes and what the problem was.

burqen commented:
@tinwelint Updated the description. Does it make sense?

tinwelint commented:
Here's my take on why this solution isn't the "correct" one:

- Injecting Runnable (even some other interface that does something similar, doesn't matter) like this is somewhat hacky, mixes responsibilities into different layers and uses abstraction to hide details that should really be made explicit on correct levels.
- From its inception, property loading inside kernel API has been eager and very simple to reason about because of that. When getAllProperties was implemented this was broken, by mistake I think. A lazy property cursor is exposed through the kernel API just so that the core API can take it and delegate it back into a method making it eager (converting it into a Map).

My proposed solution, which I discussed with @burqen is that kernel API should expose eager nodeGetAllProperties/relationshipGetAllProperties which returns Map directly and then, like magic, this whole problem goes away. So instead of working around the problem, in this case we can just move code to its right place.

davidegrohmann commented:
@tinwelint The lazy fetching properties was done on purpose and explicitly designed that way and that has _not_ being introduced by mistake. Making the loading of properties eager again is moving us backwards not forwards. I am happy to discuss a better way of implementing this fix, but I am afraid we can simply rely on eagerness.

tinwelint commented:
@davidegrohmann the property loading is eager, it was just lazy in a small and insignificant part of the execution stack to no benefit. Why should it be lazy?

MishaDemianenko commented:
@burqen @tinwelint @davidegrohmann 
Fully agree with @tinwelint: implementation is a hack and workaround, if we can remove lazy property cursors at all that should be a way to go, my believe that cursors that do low level things should do just that and transaction/statement should do what ever they do and we need to try to avoid mixing responsibilitites

MishaDemianenko reviewed (changes requested):
@burqen @tinwelint @davidegrohmann
Fully agree with @tinwelint: implementation is a hack and workaround, if we can remove lazy property cursors at all that should be a way to go, my believe that cursors that do low level things should do just that and transaction/statement should do what ever they do and we need to try to avoid mixing responsibilities

davidegrohmann commented:
@MishaDemianenko we only use lazy cursors to load properties, that has been the case since 2.3. What we expose through the OperationsFacade are "eagerized" operations on the property cursors.

tinwelint commented:
To clarify my view: Lazy cursors _inside_ the kernel API is fine and should definitely be kept, that was not in question. We just have to make sure that we don't leak the lazyness _outside_ the kernel API.

burqen commented:
@davidegrohmann @tinwelint @MishaDemianenko 
The problem is that `nodeGetProperties` and `relationshipGetProperties` is _not_ "eagerized" in OperationsFacade, but rather one level above in `NodeProxy` which makes it impossible for to use the `finally { statement.assertOpen }` solution that we use in other places in `OperationsFacade`.

EDIT: I'm thinking of 3.2. NVM

tinwelint commented:
@burqen yup, that is exactly what I'd like us to change

burqen commented:
@tinwelint Although, that solution will not work for 3.1, because NodeProxy gets property cursor from node cursor and not through OperationsFacade.

tinwelint commented:
@burqen uhm, we'll of course change NodeProxy together with OperationsFacade, won't we?

On the other hand, it would be bad to break procedures using kernel API in 3.1 and 3.2, so let's keep the solution as is and we can fix this in a later version to not let this change spiral out of control with amount of "other" work.

Commit:
Use interface 'AssertOpen' instead of Runnable to assert transaction is open

This makes the 'assertOpen' more explicit by avoiding the indirection of a runnable.

Commit:
Formatting

tinwelint commented:
@MishaDemianenko any pending comments?

MishaDemianenko commented:
@tinwelint was never pinged back have no idea what happening in here, @burqen is it ready to review round?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/AssertOpen.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    void assertOpen();
+
+    AssertOpen ALWAYS_OPEN = AlwaysOpen.INSTANCE;
```
why not just:
`    AssertOpen ALWAYS_OPEN = () ->
     {
        //always open
    };
` instead of enum that can be even additional referenced and pollute interface since can be references as `AssertOpen.AlwaysOpen.INSTANCE` without any visible in a glance reason?

tinwelint replied:
I agree, this is a weird way of doing it

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/AssertOpen.java:
```diff
@@ -0,0 +1,46 @@
+     * Assert that source tied to instance is open.
+     * Should throw exception with reason if source is not open.
+     */
+    void assertOpen();
```
if this thing should throw some exception would it make sense to throw some specific exception when underlying resource is closed (it can be runtime to avoid exception declaration anywhere), wdyt?

tinwelint replied:
That is probably what that sentence means.

MishaDemianenko replied:
what? comment is about having some specific type of exception no? comment do not mention anything about that

tinwelint replied:
If we rely on specific exceptions thrown then absolutely that should be mentioned. @burqen do we?

burqen replied:
`KernelStatement` will through two different variants of `RuntimeException`.
As that is an unchecked exception I would not put it in method signature, but maybe clarify it in javadoc. Sounds good?

From `KernelStatement`:
```
    @Override
    public void assertOpen()
    {
        if ( referenceCount == 0 )
        {
            throw new NotInTransactionException( "The statement has been closed." );
        }

        Status terminationReason = transaction.getReasonIfTerminated();
        if ( terminationReason != null )
        {
            throw new TransactionTerminatedException( terminationReason );
        }
    }
```

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -187,27 +194,28 @@ private Lock shortLivedReadLock()
     {
         return nodeRelationshipCursor.get().init( nodeRecord.isDense(), nodeRecord.getNextRel(), nodeRecord.getId(),
-                direction, relTypes );
+                direction, assertOpen, relTypes );
```
shouldn't it be the last param for the god of symmetry?)

tinwelint replied:
Because relTypes is vararg, no?

MishaDemianenko replied:
yes make sence

MishaDemianenko replied:
yes make sense

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java:
```diff
@@ -68,19 +69,21 @@
      * to place the cursor over the first item and then more calls to move the cursor through the selection.
      *
      * @param nodeId id of node to get cursor for.
+     * @param assertOpen to check if source transaction is still open on each property value fetch
```
why do we need to say that assertOpen checking source transaction? shouldn't comment have a more generic expectation from what assert open doing? (this comment related to all new assertOpen params introduced)

tinwelint replied:
May I say that some comments are simply too detailed/bike-shed:y. Would you agree?

MishaDemianenko replied:
i think anyone are free to leave comment if they think it can make PR better in one way or another

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -147,7 +147,7 @@ public StateHandlingStatementOperations(
     public Cursor<NodeItem> nodeCursor( KernelStatement statement, long nodeId )
     {
-        Cursor<NodeItem> cursor = statement.getStoreStatement().acquireSingleNodeCursor( nodeId );
+        Cursor<NodeItem> cursor = statement.getStoreStatement().acquireSingleNodeCursor( nodeId, statement );
```
using statements `assertOpen` will do volatile read each time when we read something from cursor, did we check what effect this will have on performance?

tinwelint replied:
That would be interesting yeah. I'd expect a very small (unnoticable) impact since the volatile `open` flag was checked once per property request (one or more properties) previously, now it's once per property.

burqen replied:
I have made no such investigation. Will trigger a jmh-run.
EDIT: ... once that service is up and running again.

chrisvest replied:
What we can expect to see is that the cost of a _volatile_ load compared to a normal load, is usually negligible on x86_64 except in tight loops, and slightly more expensive on Power.

burqen replied:
@chrisvest Would you go so far as to say that this is safe enough to merge without benchmarking it?

MishaDemianenko replied:
I'm more interested in the influence of this change to set of optimisations that runtime can apply. Anyway JMH run will show if this change brings any changes.

MishaDemianenko replied:
@burqen do you have any news about benchmarks run?

MishaDemianenko reviewed (commented):
@burqen couple of comments

Commit:
Cleanup weird no-op implementation in AssertOpen

Commit:
Better javadoc
